dot/parachain/collator-protocol/validator_side.go

```diff
@@ -440,21 +457,106 @@ func (cpvs CollatorProtocolValidatorSide) processMessage(msg interface{}) error
+
+		candidateHashV, err := candidateReceipt.ToPlain().Hash()
+		if err != nil {
+			return fmt.Errorf("getting candidate hash from receipt: %w", err)
+		}
+		fetchedCollation := fetchedCollationInfo{
+			relayParent:   receipt.Descriptor.RelayParent,
+			paraID:        parachaintypes.ParaID(receipt.Descriptor.ParaID),
+			candidateHash: parachaintypes.CandidateHash{Value: candidateHashV},
+			collatorID:    receipt.Descriptor.Collator,
+		}
+		// remove the candidate from the list of fetched candidates
+		collationEvent, ok := cpvs.fetchedCandidates[fetchedCollation.String()]
+		if !ok {
+			logger.Error("collation has been seconded, but the relay parent is deactivated")
+			return nil
+		}
+
+		delete(cpvs.fetchedCandidates, fetchedCollation.String())
+
+		// notify good collation
+		peerID, ok := cpvs.getPeerIDFromCollatorID(collationEvent.CollatorId)
+		if !ok {
+			return ErrPeerIDNotFoundForCollator
+		}
+		cpvs.net.ReportPeer(peerset.ReputationChange{
+			Value:  peerset.BenefitNotifyGoodValue,
+			Reason: peerset.BenefitNotifyGoodReason,
+		}, peerID)
+
+		// notify collation seconded
+		_, ok = cpvs.peerData[peerID]
+		if ok {
+			collatorProtocolMessage := NewCollatorProtocolMessage()
+			err = collatorProtocolMessage.Set(CollationSeconded{
+				RelayParent: msg.Parent,
+				Statement:   msg.Stmt,
+			})
+			if err != nil {
+				return fmt.Errorf("setting collation seconded: %w", err)
+			}
+			collationMessage := NewCollationProtocol()
+
+			err = collationMessage.Set(collatorProtocolMessage)
+			if err != nil {
+				return fmt.Errorf("setting collation message: %w", err)
+			}
 
+			err = cpvs.net.SendMessage(peerID, &collationMessage)
+			if err != nil {
+				return fmt.Errorf("sending collation message: %w", err)
+			}
+
+			perRelayParent, ok := cpvs.perRelayParent[msg.Parent]
+			if ok {
+				perRelayParent.collations.status = Seconded
+				perRelayParent.collations.secondedCount++
+				cpvs.perRelayParent[msg.Parent] = perRelayParent
+			}
+
+			// TODO: Few more things for async backing, but we don't have async backing yet
+			// https://github.com/paritytech/polkadot-sdk/blob/7035034710ecb9c6a786284e5f771364c520598d/polkadot/node/network/collator-protocol/src/validator_side/mod.rs#L1531-L1532
+		}
 	case Backed:
 		// TODO: handle backed message https://github.com/ChainSafe/gossamer/issues/3517
 	case InvalidOverseerMsg:
 		invalidOverseerMsg := msg
 
-		collationEvent, ok := cpvs.pendingCandidates[invalidOverseerMsg.Parent]
+		candidateHashV, err := msg.CandidateReceipt.Hash()
+		if err != nil {
+			return fmt.Errorf("getting candidate hash from receipt: %w", err)
+		}
+		fetchedCollation := fetchedCollationInfo{
+			relayParent:   msg.CandidateReceipt.Descriptor.RelayParent,
+			paraID:        parachaintypes.ParaID(msg.CandidateReceipt.Descriptor.ParaID),
+			candidateHash: parachaintypes.CandidateHash{Value: candidateHashV},
+			collatorID:    msg.CandidateReceipt.Descriptor.Collator,
+		}
```

Comment:
I would suggest creating a function `newFetchedCollationInfo` that receives the `candidateReceipt` and returns an instance of `fetchedCollationInfo`.
```suggestion
		fetchedCollation := newFetchedCollationInfo(msg.CandidateReceipt)
```
Same suggestion for the `case SecondedOverseerMsg:` line 479